OpenActive.Server.NET/CustomBookingEngine/CustomBookingEngine.cs

```diff
@@ -343,11 +344,19 @@ private SellerIdComponents GetSellerIdComponentsFromApiKey(Uri sellerId)
 
         public ResponseContent DeleteOrder(string clientId, Uri sellerId, string uuid)
         {
-            ProcessOrderDeletion(new OrderIdComponents { ClientId = clientId, OrderType = OrderType.Order, uuid = uuid }, GetSellerIdComponentsFromApiKey(sellerId));
-            return ResponseContent.OpenBookingNoContentResponse();
+            var result = ProcessOrderDeletion(new OrderIdComponents { ClientId = clientId, OrderType = OrderType.Order, uuid = uuid }, GetSellerIdComponentsFromApiKey(sellerId));
+            switch (result)
+            {
+                case DeleteOrderResult.OrderSuccessfullyDeleted:
+                    return ResponseContent.OpenBookingNoContentResponse();
+                case DeleteOrderResult.OrderDidNotExist:
+                    throw new OpenBookingException(new NotFoundError());
+                default:
+                    throw new Exception($"Unexpected DeleteOrderResult: {result}");
```

Comment:
The type checker asked that I include this `default` case even though all the enum cases had already been handled. I'm not sure why. Maybe because an enum value could also be null?

Reply:
Generally it's not good practice to throw a base `Exception`. Always throw the most specific class - in this case `new OpenBookingException(new OpenBookingError(), $"Unexpected DeleteOrderResult: {result}");`

Reply:
It's also worth noting that if an Exception gets thrown that is not caught and rethrown as an OpenBookingException it will return a `500` and break the Open Booking API contract, so we should watch out for that

Reply:
Why is `new OpenBookingException(new OpenBookingError(),` a more specific kind of exception? That exception is used to create error responses which are sent to the user, right e.g. "payment amount is not enough"? I was thinking that if this default case was actually hit, it would be an "unexpected internal error" as the fact of that enum case not being handled is entirely an internal issue. Therefore, a 500 might be appropriate

Reply:
Ideally, the type system would just catch this - then we wouldn't have to have this conversation. And the type system would throw an error if an additional option was added to the enum

Reply:
There is this error class: InvalidEnumArgumentException (https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.invalidenumargumentexception?view=netcore-3.1). Maybe that's the one?

Reply:
Could make a custom exception class for "invalid enum values"?

Reply:
Or maybe there's an equivalent data type that is not backed by a numeric value and can therefore only be one of the selected options..?

Reply:
> It's also worth noting that if an Exception gets thrown that is not caught and rethrown as an OpenBookingException it will return a 500 and break the Open Booking API contract, so we should watch out for that

Are 500s (and 503s, etc) not always a possible response from a server in case something has gone fatally wrong?

Reply:
Yes agree ideally this would be handled by the type system, but alas C# does have this kind of challenge...

I guess the question is more broadly: within our library, what type of Exception should we throw when something unexpected happens. I believe OpenBookingException defaults to a 500, so `OpenBookingError` would indeed throw a 500.

As an aside: we probably want to go through the project at some point and search for all instances of OpenBookingError to see if a more specific subclass can be defined and added to the booking spec.

The challenge with not throwing an OpenBookingException here is that ASP.NET won't spit out JSON (the framework renders an HTML 500 page for uncaught exceptions).

Another option is that we define another type of exception as you say, perhaps alongside the "EngineConfigurationException", or a subclass of "EngineConfigurationException", and then update the [controllers](https://github.com/openactive/OpenActive.Server.NET/blob/master/Examples/BookingSystem.AspNetCore/Controllers/OpenBookingController.cs) to catch this.

The advantage of this is these errors can be logged by the application and handled differently to `OpenBookingError` expected exceptions. We could also implement `obe.ErrorResponseContent.GetContentResult()` on `EngineConfigurationException` too.

Whatyathink?

Reply:
Perhaps the booking spec should just include `InternalApplicationError`, `InternalLibraryError` and `InternalLibraryConfigurationError` as a subclass of `OpenBookingError`. This is likely the better place to address the issue, as such a subclass could be used across all implementations / languages (where all will likely have the same issue)

Reply:
Okay thanks Nick, that helps me understand it greatly. So I'll set it to an OpenBookingError for now - so that the framework registers it as a sanctioned unexpected error and doesn't return an HTML page.

What you've suggested as an alternative solution makes more sense to me conceptually, but it may not be worth making larger changes for this rather small issue and I think I still have more to learn before attempting any wider changes

Reply:
But this exchange has been very helpful to me!

Reply:
Ok great! Will add InternalApplicationError, InternalLibraryError and InternalLibraryConfigurationError as a subclass of OpenBookingError to the Open Booking API spec backlog for future consideration